Allow negative quantity for product stock

tswfi opened the pull request:
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Product stock quantity should be allowed to be negative. Adapter wrongly always got the abs value when saving combinations in BO
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #25545
| How to test?      | Save negative value to product combination stocklevel, check that the value is negative in ps_stock_available table also. reload the page, check that the value is still negative in the combination edit list.
| Possible impacts? |

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/25549)
<!-- Reviewable:end -->

tswfi committed:
Allow negative quantity for product stock, fixes #25545

Progi1984 commented on src/PrestaShopBundle/Model/Product/AdminModelAdapter.php:
```diff
@@ -339,9 +339,8 @@ public function getModelData($form_data, $isMultiShopContext = false)
-            $form_data['combinations'][$k]['attribute_quantity'] = abs(
-                $this->floatParser->fromString($combination['attribute_quantity'])
-            );
+            $form_data['combinations'][$k]['attribute_quantity'] =
+                $this->floatParser->fromString($combination['attribute_quantity']);
```
```suggestion
            $form_data['combinations'][$k]['attribute_quantity'] = $this->floatParser->fromString($combination['attribute_quantity']);
```

Progi1984 reviewed (commented):
@tswfi Thank you for your contribution. I have brought to your PR a small style fix.

tswfi committed:
Add lenient parsing for negative sign to floatParser (allow all CLDR minussigns)

tswfi commented:
Thanks for checking @Progi1984 and @PierreRambaud . 
I continued testing this and the minussign issue was still present (see issue #14913)
commit 42634b2 allows all the different minusSigns to be used for floatParsing which fixes this bug for dev (but not for 1.7.7.5)
UI still has some js that converts all extra characters to `.` when editing the field (I think that is a separate issue).

hibatallahAouadni commented:
Reminder: This PR cannot be approved till we have the PM approval on the related issue

tswfi commented:
@hibatallahAouadni maybe I should have never talked about the other issue in this pr.

The thing is both of these changes are required to fix this problem in develop branch.

Maybe this video demonstration helps to see the problem in current develop branch.

https://drive.google.com/file/d/1_0BR1NN2V9gFKqgAH46YJ76FSLrc5LM0/view?usp=sharing

ps, I rebased the pr just for easier video recording :D

pps, I removed the "fixes 14913" from the description (though I think that issue can be closed if this pr gets into 1.7.8.1)

ppps, and it seems I found another bug while recording... quantity field changes from `−321` to `.322` if you change only the last number... oh, now that I checked my previous comments I already saw that before also :D I wonder if I already made an issue for it ...

jolelievre reviewed (dismissed):
That's a smart way to fix the problem, I'll try and keep that in mind for the new product page
Although I'm not sure this fix will have much impact in the next release since the product page will be replaced with the new one

jolelievre commented:
Also a side note, the quantity edition changed in the latest page Now we only edit quantity via a "delta modification" like it is done in stock This resolves a long time issue that some inconsistencies could happen when the merchant edits the quantity while it changed since he opened the product page (if an order was submitted in the meantime for example)

tswfi commented:
@Progi1984 rebased

PierreRambaud reviewed (changes requested):
Could you please add unit tests in FloatParserTest.php?

tswfi committed:
added tests

tswfi commented:
> Could you please add unit tests in FloatParserTest.php?

done in 512f97f

PierreRambaud commented on tests/Unit/PrestaShopBundle/Utils/FloatParserTest.php:
```diff
@@ -97,6 +97,29 @@ public function provideValidStrings()
         ];
+
+        // cldr minus signs
+        $cldr_minus_signs = $this->str_split_unicode('-‒⁻₋−➖﹣－');
```
Could you please only fill the array above and don't create something dynamic.
Seems you copy & paste a method. We are not using snake case anywhere :thinking:

tswfi replied:
I was thinking it is easier to update the test case when that string can be copied from cldr definition if it changes there in the future.

continued the `str_split` to have `_unicode` is where the naming convention comes from.. mb_str_split would work also but it is only available from php7.4 forward...

I'll modify the test case in a bit.

tswfi replied:
done in d56f3e8

PierreRambaud replied:
Tests have to be as simple as possible, more when it concerns data providers. (That's not something you doing when using TDD)
Filling the array will be enough.

tswfi committed:
change test case as per requested

Progi1984 commented:
The issue #25545 is always in "Waiting for @PrestaShop/product-team"

Progi1984 commented:
Ping @MatShir

tswfi commented:
Ping someone :)

matks commented:
Issue https://github.com/PrestaShop/PrestaShop/issues/25545 is not in "waiting for PM" anymore but needs specifications about expected behavior

matks commented:
@tswfi To provide some explanations: in the past we did validate too fast some changes that later confused a lot of users, which required then to revert them.

Now we try, before making changes with impact to the software, to lay down all usecases and parameters of the topic to explore them and be sure we are not shooting a bullet into our own foot 😅 

If you want to speed up the process to validate the PR maybe you can assist @MatShir or @marionf in laying down all usecases and parameters of the topic, all situations it can encounter, all configurations... to make sure the software becomes better.

tswfi commented:
The use case is that the "stocklevel" field in BO has a negative number like "-10" and currently it is saved, depending on which bug triggers, as +10 (`abs` bug) or 0 (CLDR bug).

If a merchant writes a value to a field it should either save that value or give a definitive warning and do nothing.

Negative stock amount is a valid situation (the cart does not validate quantities before allowing the order through in all steps, but that is a another issue and a pr).

Negative stock level is also valid when doing "presales" for products.

Hlavtox commented:
@matks
If customer buys 5 pieces of product out of stock product, quantity is -5.

You should be definitely able to modify it to a negative value, which is perfectly valid.

jolelievre commented:
This PR targets the legacy page, so I don't think it's worth investing time on this page nor its related code since they will be removed before the next release anyway

But know that on the new page it's possible to set a negative stock, we now use a delta quantity to edit the stock (like in the stock page), and it allows negative values even when the end result will cause a negative stock. Feel free to test the experimental page to test this feature.

Hlavtox commented:
OK, perfect!!! 👍

MatShir commented:
Hi guys, 
We placed your issue about 'minus' in our kanban to fix it when the combination part would be fully migrated https://github.com/PrestaShop/PrestaShop/issues/26502 ;)
If you check the develop branch, the 'add a negative' feature is working.
Sorry for the inconveniences
